### PR TITLE
Interruption tests

### PR DESCRIPTION
**Pull request by Ostrokrzew**
Interruption tests:
- core flushing interruption;
- cache flushing interruption
- core removal interruption
- switching cache mode interruption
- cache stopping interruption

Signed-off-by: Slawomir Jankowski slawomir.jankowski@intel.com

### CODE REVIEW

File: test/functional/tests/fault_injection/test_fault_injection_interrupts.py [hunk 1]
```diff
@@ -0,0 +1,534 @@
...
+@pytest.mark.require_disk("core", DiskTypeLowerThan("cache"))
+def test_interrupt_cache_mode_switch_immediately(cache_mode, filesystem):
+    """
+        title: Test if cache mode switching works correctly.
+        description: |
+          Negative test of the ability of OpenCAS to handle cache mode switching
+          immediate interruption.
+        pass_criteria:
+          - No system crash.
+          - Cache mode will not be switched after interruption.
```
Comment (mmichal10) [inline test/functional/tests/fault_injection/test_fault_injection_interrupts.py]: you don't check if cache mode left unchanged

File: test/functional/tests/fault_injection/test_fault_injection_interrupts.py [hunk 2]
```diff
@@ -0,0 +1,534 @@
...
+    """
+        title: Test if cache mode switching works correctly.
+        description: |
+          Negative test of the ability of OpenCAS to handle cache mode switching
+           interruption with delay.
+        pass_criteria:
+          - No system crash.
+          - Cache mode cannot be interrupted with delay.
+          - Md5sum are correct during all test steps.
+          - Dirty blocks quantity after ache mode switching is zero.
```
Comment (mmichal10) [inline test/functional/tests/fault_injection/test_fault_injection_interrupts.py]: `ache mode`

File: test/functional/tests/fault_injection/test_fault_injection_interrupts.py [hunk 3]
```diff
@@ -0,0 +1,520 @@
...
+@pytest.mark.require_disk("cache", DiskTypeSet([DiskType.optane, DiskType.nand]))
+@pytest.mark.require_disk("core", DiskTypeLowerThan("cache"))
+def test_interrupt_core_flush(cache_mode, filesystem):
+    """
+        title: Test if core's flushing works correctly.
+        description: |
+          Negative test of the ability of OpenCAS to handle core flushing interruption.
+        pass_criteria:
+          - No system crash.
+          - Cache's metadata on core device will not be flushed after interruption.
```
[the remark this answers is not in the document]
Comment (Ostrokrzew) [reply test/functional/tests/fault_injection/test_fault_injection_interrupts.py]: rephrased

File: test/functional/tests/fault_injection/test_fault_injection_interrupts.py [hunk 4]
```diff
@@ -274,3 +274,159 @@ def test_interrupt_core_remove(cache_mode, filesystem):
...
+                TestRun.LOGGER.error("Cache mode should remain the same.")
+
+        with TestRun.step("Flush cache."):
+            cache.flush_cache()
+
+        with TestRun.step("Check md5 sum of test file again."):
+            if tf_md5s_bf != test_file.md5sum():
+                TestRun.LOGGER.error(
+                    "Md5 sums before and after interrupting core removal are different.")
+
```
Comment (mmichal10) [inline test/functional/tests/fault_injection/test_fault_injection_interrupts.py]: Test would be more valuable if cache was stopped, and md5 was checked on core disk. Currently even if all data is clean, it is still cached, so you don't check if file was properly written to core.

File: test/functional/tests/fault_injection/test_fault_injection_interrupts.py [hunk 5]
```diff
@@ -0,0 +1,542 @@
...
+            if cache.get_cache_mode() == cache_mode:
+                TestRun.LOGGER.error("Cache mode should have changed.")
+
+        with TestRun.step("Flush cache."):
+            cache.flush_cache()
+
+        with TestRun.step("Check md5 sum of test file again."):
+            if tf_md5s_bf != test_file.md5sum():
+                TestRun.LOGGER.error(
+                    "Md5 sums before and after interrupting core removal are different.")
```
Comment (mmichal10) [inline test/functional/tests/fault_injection/test_fault_injection_interrupts.py]: Test would be more valuable if cache was stopped, and md5 was checked on core disk. Currently even if all data is clean, it is still cached, so you don't check if file was properly written to core.

File: test/functional/tests/fault_injection/test_fault_injection_interrupts.py [hunk 6]
```diff
@@ -0,0 +1,542 @@
...
+        with TestRun.step("Mount core device."):
+            core_part.mount(mount_point)
+
+        with TestRun.step("Flush cache."):
+            cache.flush_cache()
+
+        with TestRun.step("Check md5 sum of test file again."):
+            if tf_md5s_bf != test_file.md5sum():
+                TestRun.LOGGER.error(
+                    "Md5 sums before and after interrupting core removal are different.")
```
Comment (mmichal10) [inline test/functional/tests/fault_injection/test_fault_injection_interrupts.py]: Test would be more valuable if cache was stopped, and md5 was checked on core disk. Currently even if all data is clean, it is still cached, so you don't check if file was properly written to core.

File: test/functional/tests/fault_injection/test_fault_injection_interrupts.py [hunk 7]
```diff
@@ -0,0 +1,542 @@
...
+        with TestRun.step("Mount core."):
+            core.mount(mount_point)
+
+        with TestRun.step("Flush cache."):
+            cache.flush_cache()
+
+        with TestRun.step("Check md5 sum of test file again."):
+            if tf_md5s_bf != test_file.md5sum():
+                TestRun.LOGGER.error("Md5 sums before and after interrupting"
+                                     " cache stop are different.")
```
Comment (mmichal10) [inline test/functional/tests/fault_injection/test_fault_injection_interrupts.py]: Test would be more valuable if cache was stopped, and md5 was checked on core disk. Currently even if all data is clean, it is still cached, so you don't check if file was properly written to core.

File: test/functional/tests/fault_injection/test_fault_injection_interrupts.py [hunk 8]
```diff
@@ -99,3 +99,82 @@ def test_interrupt_core_flush(cache_mode, filesystem):
...
+        with TestRun.step("Check number of dirty data on exported object after interruption."):
+            cache_drtblcks_after = cache.get_dirty_blocks()
+            if cache_drtblcks_after >= cache_drtblcks_before:
+                TestRun.LOGGER.error("Quantity of dirty lines after cache flush interruption "
+                                     "should be lower.")
+            if int(cache_drtblcks_after) == 0:
+                TestRun.LOGGER.error("Quantity of dirty lines after cache flush interruption "
+                                     "should not be zero.")
+
+        with TestRun.step("Stop cache."):
```
Comment (mmichal10) [inline test/functional/tests/fault_injection/test_fault_injection_interrupts.py]: Exported object is still mounted, you cant stop cache.

File: test/functional/tests/fault_injection/test_fault_injection_interrupts.py [hunk 9]
```diff
@@ -0,0 +1,536 @@
...
+            core_drtblcks_after = core.get_dirty_blocks()
+            if core_drtblcks_after >= core_drtblcks_before:
+                TestRun.LOGGER.error("Quantity of dirty lines after core flush interruption "
+                                     "should be lower.")
+            if int(core_drtblcks_after) == 0:
+                TestRun.LOGGER.error("Quantity of dirty lines after core flush interruption "
+                                     "should not be zero.")
+
+        with TestRun.step("Stop cache."):
+            casadm.stop_all_caches()
```
Comment (mmichal10) [inline test/functional/tests/fault_injection/test_fault_injection_interrupts.py]: Exported object is still mounted, you cant stop cache.

File: test/functional/tests/fault_injection/test_fault_injection_interrupts.py [hunk 10]
```diff
@@ -0,0 +1,536 @@
...
+                                     "should be lower.")
+            if int(cache_drtblcks_after) == 0:
+                TestRun.LOGGER.error("Quantity of dirty lines after cache flush interruption "
+                                     "should not be zero.")
+
+        with TestRun.step("Check cache mode."):
+            if cache.get_cache_mode() != cache_mode:
+                TestRun.LOGGER.error("Cache mode should remain the same.")
+
+        with TestRun.step("Stop cache."):
```
Comment (mmichal10) [inline test/functional/tests/fault_injection/test_fault_injection_interrupts.py]: Exported object is still mounted, you cant stop cache.

File: test/functional/tests/fault_injection/test_fault_injection_interrupts.py [hunk 11]
```diff
@@ -0,0 +1,536 @@
...
+                "Get quantity of dirty data on exported object after sending interruption "
+                "signal to cas to stop mode switching."):
+            if int(cache.get_dirty_blocks()) != 0:
+                TestRun.LOGGER.error("Quantity of dirty lines should be zero now.")
+
+        with TestRun.step("Check cache mode."):
+            if cache.get_cache_mode() == cache_mode:
+                TestRun.LOGGER.error("Cache mode should have changed.")
+
+        with TestRun.step("Stop cache."):
```
Comment (mmichal10) [inline test/functional/tests/fault_injection/test_fault_injection_interrupts.py]: Exported object is still mounted, you cant stop cache.

File: test/functional/tests/fault_injection/test_fault_injection_interrupts.py [hunk 12]
```diff
@@ -0,0 +1,536 @@
...
+        with TestRun.step("Check number of dirty data on exported object after interruption."):
+            cache_drtblcks_after = cache.get_dirty_blocks()
+            if cache_drtblcks_after >= cache_drtblcks_before:
+                TestRun.LOGGER.error("Quantity of dirty lines after cache stop interruption "
+                                     "should be lower.")
+            if int(cache_drtblcks_after) == 0:
+                TestRun.LOGGER.error("Quantity of dirty lines after cache stop interruption "
+                                     "should not be zero.")
+
+        with TestRun.step("Stop cache."):
```
Comment (mmichal10) [inline test/functional/tests/fault_injection/test_fault_injection_interrupts.py]: Exported object is still mounted, you cant stop cache.

File: test/functional/tests/fault_injection/test_fault_injection_interrupts.py [hunk 13]
```diff
@@ -0,0 +1,540 @@
...
+            core_part.unmount()
+
+
+@pytest.mark.parametrize("filesystem", Filesystem)
+@pytest.mark.parametrize("cache_mode", CacheMode.with_traits(CacheModeTrait.LazyWrites))
+@pytest.mark.require_disk("cache", DiskTypeSet([DiskType.optane, DiskType.nand]))
+@pytest.mark.require_disk("core", DiskTypeLowerThan("cache"))
+def test_interrupt_cache_mode_switch_immediately(cache_mode, filesystem):
+    """
+        title: Test if cache mode switching works correctly.
```
Comment (katlapinka) [inline test/functional/tests/fault_injection/test_fault_injection_interrupts.py]: after interruption

File: test/functional/tests/fault_injection/test_fault_injection_interrupts.py [hunk 14]
```diff
@@ -0,0 +1,540 @@
...
+            core_part.unmount()
+
+
+@pytest.mark.parametrize("filesystem", Filesystem)
+@pytest.mark.parametrize("cache_mode", CacheMode.with_traits(CacheModeTrait.LazyWrites))
+@pytest.mark.require_disk("cache", DiskTypeSet([DiskType.optane, DiskType.nand]))
+@pytest.mark.require_disk("core", DiskTypeLowerThan("cache"))
+def test_interrupt_cache_mode_switch_delayed(cache_mode, filesystem):
+    """
+        title: Test if cache mode switching works correctly.
```
Comment (katlapinka) [inline test/functional/tests/fault_injection/test_fault_injection_interrupts.py]: This test has the same tittle as the above test - it can't be like that

**Pull request merged**